File: python/tests/algorithmic_mix_test.py
import time, math, random, json, os

class AlgorithmicMixTestTest:
    def __init__(self, size=5000, depth=8, runs=5):
        self.name = "Algorithmic Mix"
        self.size = size
        self.depth = depth
        self.runs = runs

    def _recursive_fib(self, n):
        if n < 2:
            return n
        return self._recursive_fib(n - 1) + self._recursive_fib(n - 2)

    def _sort_and_search(self, arr):
        arr.sort()
        target = arr[len(arr)//2]
        return arr.index(target)

    def _io_cycle(self, data):
        fname = "temp_mix_test.tmp"
        with open(fname, "w") as f:
            json.dump(data, f)
        with open(fname, "r") as f:
            return json.load(f)
        os.remove(fname)

    def run_once(self):
        numbers = [random.randint(1, 100000) for _ in range(min(self.size, 5000))]  # ✅ Cap to safe size

        t0 = time.perf_counter()
        math_sum = sum(math.sin(i) * math.sqrt(i % 100 + 1) for i in range(1, 500))
        fib_val = self._recursive_fib(10)
        idx = self._sort_and_search(numbers)
        _ = self._io_cycle({"sum": math_sum, "fib": fib_val, "idx": idx, "nums": numbers[:50]})
        t1 = time.perf_counter()

        duration = t1 - t0
        return {"duration_s": duration, "ops_per_s": 1 / duration if duration > 0 else 0}

    def run(self, runs=None, iterations=None):
        if runs is not None:
            self.runs = runs
        if iterations is not None:
            self.size = min(iterations, 5000)  # ✅ Avoid huge freeze

        results = [self.run_once() for _ in range(self.runs)]
        times = sorted(r["duration_s"] for r in results)
        ops = sorted(r["ops_per_s"] for r in results)
        return {
            "name": self.name,
            "runs": self.runs,
            "median_time_s": times[len(times)//2],
            "median_ops_per_s": ops[len(ops)//2],
            "raw": results
        }

if __name__ == "__main__":
    result = AlgorithmicMixTestTest().run()
    with open("results_algorithmic_mix_python.json", "w") as f:
        json.dump(result, f, indent=2)
    print(json.dumps(result, indent=2))

File: ruby/tests/algorithmic_mix_test.rb
require "json"
require "time"
require "securerandom"

class AlgorithmicMix
  attr_reader :name

  def initialize(size = 5000, depth = 8, runs = 5)
    @name = "Algorithmic Mix"
    @size = size
    @depth = depth
    @runs = runs
  end

  # Recursive Fibonacci (small depth for realism)
  def recursive_fib(n)
    return n if n < 2
    recursive_fib(n - 1) + recursive_fib(n - 2)
  end

  # Sorting and searching test
  def sort_and_search(arr)
    arr.sort!
    target = arr[arr.length / 2]
    arr.index(target)
  end

  # File I/O cycle test with a unique temp file name
  def io_cycle(data)
    fname = "temp_mix_test_#{SecureRandom.hex(4)}.tmp"
    File.open(fname, "w") { |f| f.write(JSON.dump(data)) }
    parsed = JSON.parse(File.read(fname))
    File.delete(fname) if File.exist?(fname)
    parsed
  end

  # One full run (includes arithmetic, recursion, sort/search, file I/O)
  def run_once
    numbers = Array.new(@size) { rand(1..100_000) }

    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    # Arithmetic + recursion
    math_sum = (1..500).reduce(0.0) { |acc, i| acc + Math.sin(i) * Math.sqrt((i % 100) + 1) }
    fib_val = recursive_fib(10)

    # Sorting + searching
    idx = sort_and_search(numbers)

    # File I/O
    data = { "sum" => math_sum, "fib" => fib_val, "idx" => idx, "nums" => numbers[0..50] }
    _ = io_cycle(data)

    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    duration = t1 - t0

    { "duration_s" => duration, "ops_per_s" => 1.0 / (duration > 0 ? duration : 1e-9) }
  end

  # Main test runner (now compatible with automated harness)
  def run(runs = 5, iterations = nil)
    runs ||= @runs
    iterations ||= 1  # not used, but included for compatibility

    results = Array.new(runs) { run_once }
    times = results.map { |r| r["duration_s"] }.sort
    ops = results.map { |r| r["ops_per_s"] }.sort
    median_t = times[times.size / 2]
    median_ops = ops[ops.size / 2]

    {
      "name" => @name,
      "runs" => runs,
      "median_time_s" => median_t,
      "median_ops_per_s" => median_ops,
      "raw" => results
    }
  end
end

# Alias for test harness compatibility
AlgorithmicMixTestTest = AlgorithmicMix

# Execute directly if script is run standalone
if __FILE__ == $0
  result = AlgorithmicMix.new.run
  File.write("results_algorithmic_mix_ruby.json", JSON.pretty_generate(result))
  puts JSON.pretty_generate(result)
end
